tests/test_utils.py

```diff
@@ -134,6 +136,26 @@ def test_patch_environment_key_exists(self):
         assert "BB" not in os.environ
         assert "CC" not in os.environ
 
+    def test_patch_environment_restores_on_error(self):
+        # we need to find an upper-case envvar
+        # because `patch_environment upper-cases all keys...
+        key, orig_value = next(kv for kv in os.environ.items() if kv[0].isupper())
+        new_value = orig_value[::-1]
+        with pytest.raises(RuntimeError), patch_environment(**{key: new_value}):
+            assert os.environ[key] == os.getenv(key) == new_value  # noqa: TID251
+            raise RuntimeError("Oopsy daisy!")
+        assert os.environ[key] == os.getenv(key) == orig_value  # noqa: TID251
```

Comment:
Why test both `os.environ[key]` and `os.getenv(key)`, can they differ?

Reply:
They can't, but the issue with the preimage of `clear_environment()` was exactly that it wouldn't touch what `os.getenv()` would return, so it's belt-and-suspenders here :)